libyul/ASTLabelRegistry.cpp

```diff
@@ -0,0 +1,172 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/ASTLabelRegistry.h>
+
+#include <libyul/Exceptions.h>
+
+#include <fmt/format.h>
+
+#include <range/v3/algorithm/all_of.hpp>
+#include <range/v3/algorithm/max.hpp>
+#include <range/v3/view/map.hpp>
+
+using namespace solidity::yul;
+
+
+ASTLabelRegistry::ASTLabelRegistry(): m_labels{""}, m_idToLabelMapping{0} {}
+
+ASTLabelRegistry::ASTLabelRegistry(std::vector<std::string> _labels, std::vector<size_t> _idToLabelMapping)
+{
+	yulAssert(!_labels.empty());
+	yulAssert(_labels[0].empty());
+	yulAssert(!_idToLabelMapping.empty());
+	yulAssert(_idToLabelMapping[0] == 0);
+	// using vector<uint8_t> over vector<bool>, as the latter is optimized for space-efficiency
+	std::vector<uint8_t> labelVisited (_labels.size(), false);
+	size_t numLabels = 0;
+	for (auto const& labelIndex: _idToLabelMapping)
+	{
+		if (labelIndex == ghostLabelIndex())
+			continue;
+		yulAssert(labelIndex < _labels.size());
+		// it is possible to have multiple references to the empty / ghost label index
+		// only the id == 0 refers to an actually empty label, otherwise the LabelID is unused
+		yulAssert(
+			labelIndex == 0 || !labelVisited[labelIndex],
+			fmt::format("LabelID {} (label \"{}\") is not unique.", labelIndex, _labels[labelIndex])
+		);
+		labelVisited[labelIndex] = true;
+		if (labelIndex >= 1)
+			++numLabels;
+	}
+	yulAssert(numLabels + 1 == _labels.size(), "Unused labels present.");
```

Comment:
I'd call these labels "unreferenced" to avoid confusion with unused IDs:

```suggestion
	yulAssert(numLabels + 1 == _labels.size(), "Unreferenced labels present.");
```